backupresolver: avoid fetching all descriptors

rafiss opened the pull request:
```
goos: darwin
goarch: arm64
cpu: Apple M1 Pro
BenchmarkResolveTargets

BenchmarkResolveTargets/OldImplementation         	      91	  18089891 ns/op	17230891 B/op	  109217 allocs/op

BenchmarkResolveTargets/NewImplementation 	              1232	    846169 ns/op	  282877 B/op	    2225 allocs/op
```

Please review individual commits.

---

### backupresolver: avoid fetching all descriptors

This adds a new API for use by BACKUP code to resolve TablePatterns to
descriptors that need to be backed up.

Unlike the existing API, this avoids fetching all descriptors that exist
in the database. Instead, the returned list will only contain the subgraph
of all descriptors involved in the backup:

- Table backup: Includes the table descriptors, their parent schema descriptor,
      and parent database descriptor. Also includes any immutable descriptors
      that the table depends like UDTs.
- Database backup: Includes the database descriptor, and all children (table, schema,
      sequences, udfs, etc)

### backup,changefeedccl: use new ResolveTargets API

This updates the production code paths to use the new
function that was introduced in the previous commit.

As part of this, the function was enhanced to match the extra return
parameters of ResolveTargetsToDescriptors.

### backupresolver: remove ResolveTargetsToDescriptors

### backupresolver: fix lookup logic for descriptors

The new implementation of ResolveTargets had an overly aggressive
filter. We need to include these descriptors in order for backups taken
during a schema change to include the correct descriptors. This was need
to make TestBackupSuccess_base_create_index_create_schema_separate_statements
pass reliably.

This also fixes the error handling when looking up schemas to handle
synthetic public schemas correctly. In addition, the resolution logic
now skips temporary objects.

Lastly, the name resolver logic previously was looking up types. In the
context of resolving BACKUP targets, we only ever need to resolve tables
by name, and the type fallback logic was making the wrong object get
resolved if the table name shared a name with a type.

### backupresolver: add all wildcard expansions to expandedDBs

Previously, ResolveTargets only added database IDs to expandedDBs for
database-level wildcards (db.*), but not for schema-level wildcards
(db.schema.* or schema.*). This didn't match the behavior of
DescriptorsMatchingTargets, which adds to ExpandedDB for any wildcard
pattern in the AllTablesSelector case (targets.go:543-546).

This inconsistency matters because expandedDBs is used downstream to
determine which databases had wildcard expansion applied, affecting
incremental backup behavior and descriptor change tracking.

This commit updates ResolveTargets to add to expandedDBs for all
wildcard patterns, matching the original DescriptorsMatchingTargets
behavior.

### backupresolver: adjust test message expectations

This adjusts tests to check for the appropriate errors that are returned
by the new ResolveTargets implementation.

- The old code delegated to DescriptorsMatchingTargets, which was returning
      an incorrectly specific error message when a database did not exist.
- To handle wildcard expansion, The previous implementation was calling
      ResolveObjectPrefix, which mutates the object name in place to add the
      `public` schema to the pattern if it was unspecified. This was not
      correct, as the db.* pattern causes all tables in the database to be
      backed up, not just the ones in the public schema.

---

fixes: #146803
Release note (performance improvement): Database- and table-level backups no longer fetch all object descriptors from disk in order to resolve the backup targets. Now only the objects that are referenced by the targeted objects will be fetched. This improves performance when there are many tables in the cluster.

rafiss committed:
backupresolver: avoid fetching all descriptors

This adds a new API for use by BACKUP code to resolve TablePatterns to
descriptors that need to be backed up.

Unlike the existing API, this avoids fetching all descriptors that exist
in the database. Instead, the returned list will only contain the subgraph
of all descriptors involved in the backup:

- Table backup: Includes the table descriptors, their parent schema descriptor,
  and parent database descriptor. Also includes any immutable descriptors
  that the table depends like UDTs.
- Database backup: Includes the database descriptor, and all children (table, schema,
  sequences, udfs, etc)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/157790)
<!-- Reviewable:end -->

rafiss committed:
backup,changefeedccl: use new ResolveTargets API

This updates the production code paths to use the new
function that was introduced in the previous commit.

As part of this, the function was enhanced to match the extra return
parameters of ResolveTargetsToDescriptors.

Release note (performance improvement): Database- and table-level backups
no longer fetch all object descriptors from disk in order to resolve the
backup targets. Now only the objects that are referenced by the targeted
objects will be fetched. This improves performance when there are many
tables in the cluster.

rafiss committed:
backupresolver: remove ResolveTargetsToDescriptors

Release note: None

rafiss committed:
backupresolver: fix lookup logic for descriptors

The new implementation of ResolveTargets had an overly aggressive
filter. We need to include these descriptors in order for backups taken
during a schema change to include the correct descriptors. This was need
to make TestBackupSuccess_base_create_index_create_schema_separate_statements
pass reliably.

This also fixes the error handling when looking up schemas to handle
synthetic public schemas correctly. In addition, the resolution logic
now skips temporary objects.

Lastly, the name resolver logic previously was looking up types. In the
context of resolving BACKUP targets, we only ever need to resolve tables
by name, and the type fallback logic was making the wrong object get
resolved if the table name shared a name with a type.

Release note: None

rafiss committed:
backupresolver: add all wildcard expansions to expandedDBs

Previously, ResolveTargets only added database IDs to expandedDBs for
database-level wildcards (db.*), but not for schema-level wildcards
(db.schema.* or schema.*). This didn't match the behavior of
DescriptorsMatchingTargets, which adds to ExpandedDB for any wildcard
pattern in the AllTablesSelector case (targets.go:543-546).

This inconsistency matters because expandedDBs is used downstream to
determine which databases had wildcard expansion applied, affecting
incremental backup behavior and descriptor change tracking.

This commit updates ResolveTargets to add to expandedDBs for all
wildcard patterns, matching the original DescriptorsMatchingTargets
behavior.

Release note: None

rafiss committed:
backupresolver: adjust test message expectations

This adjusts tests to check for the appropriate errors that are returned
by the new ResolveTargets implementation.

- The old code delegated to DescriptorsMatchingTargets, which was returning
  an incorrectly specific error message when a database did not exist.
- To handle wildcard expansion, The previous implementation was calling
  ResolveObjectPrefix, which mutates the object name in place to add the
  `public` schema to the pattern if it was unspecified. This was not
  correct, as the db.* pattern causes all tables in the database to be
  backed up, not just the ones in the public schema.

Release note: None

msbutler commented on pkg/backup/backupresolver/targets_optimized.go:
```diff
@@ -30,10 +30,25 @@ import (
 	ctx context.Context, p sql.PlanHookState, endTime hlc.Timestamp, targets *tree.BackupTargetList,
-) ([]catalog.Descriptor, error) {
+) (
+	[]catalog.Descriptor,
```
your plan makes sense to me!

msbutler commented on pkg/backup/backupresolver/targets.go:
```diff
@@ -674,35 +676,472 @@ func LoadAllDescs(
+			err = allInDB.ForEachDescriptor(func(desc catalog.Descriptor) error {
+				// Skip virtual tables and temp tables.
+				if table, ok := desc.(catalog.TableDescriptor); ok {
+					if desc.Dropped() || (!desc.Public() && !desc.Offline()) {
```
could this line be simplified to: ` !(desc.Public() || desc.Offline())`?

msbutler commented on pkg/backup/backupresolver/targets.go:
```diff
@@ -674,35 +676,472 @@ func LoadAllDescs(
+					return nil
+				}
+
+				// For schemas, only skip if dropped (include PUBLIC, OFFLINE, and ADD).
```
huh, weird that we include ADDing schemas but not ADDing tables

msbutler commented on pkg/backup/backupresolver/targets.go:
```diff
@@ -674,35 +676,472 @@ func LoadAllDescs(
+				}
+
+				table, ok := desc.(catalog.TableDescriptor)
+				if !ok || !table.Public() || table.IsTemporary() {
```
nit: if `!ok` is returned, returning does not exist makes sense. But for non public or temp tables, perhaps the error should be `cannot be backed up`?

msbutler commented on pkg/backup/backupresolver/targets.go:
```diff
@@ -674,35 +676,472 @@ func LoadAllDescs(
+			if err != nil {
+				return err
+			}
+			if !schema.Dropped() {
```
should the virtual and temp schema check be in here too?

msbutler commented on pkg/backup/backupresolver/targets.go:
```diff
@@ -674,35 +676,472 @@ func LoadAllDescs(
+							return nil
+						}
+						// For other descriptors (types, functions), apply standard filter.
+						if desc.Dropped() || (!desc.Public() && !desc.Offline()) {
```
nit: same comment as above

msbutler reviewed (commented):
just took a first pass at the source code. this is sooooo much cleaner! thank you so much for working on this!

msbutler reviewed (approved):
Amazing work!

msbutler commented on pkg/backup/backupresolver/targets.go:
```diff
@@ -674,35 +676,472 @@ func LoadAllDescs(
+							addDesc(table)
+							return nil
+						}
+						addDesc(desc)
```
would be nice to add an unreferenced type to the user defined schema in the unit test. i can also add it as a follow up.

rafiss replied:
i added this to the test.
